datastore/mqtt/mqtt.go

```diff
@@ -0,0 +1,208 @@
+	registerMetricsOnce(metrics)
+
+	if config.PublishTimeout == 0 {
+		config.PublishTimeout = 1 // Default to 1 second
+	}
+	if config.ConnectTimeout == 0 {
+		config.ConnectTimeout = 30 // Default to 30 seconds
+	}
+
+	opts := pahomqtt.NewClientOptions().
+		AddBroker(config.Broker).
+		SetClientID(config.ClientID).
+		SetUsername(config.Username).
+		SetPassword(config.Password).
+		SetConnectTimeout(time.Duration(config.ConnectTimeout) * time.Second)
+
+	client := PahoNewClient(opts)
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
+	return &MQTTProducer{
+		client:             client,
+		config:             config,
+		logger:             logger,
+		airbrakeHandler:    airbrakeHandler,
+		namespace:          namespace,
+		ctx:                ctx,
+		ackChan:            ackChan,
+		reliableAckTxTypes: reliableAckTxTypes,
+	}, nil
+}
+
+func (p *MQTTProducer) Produce(rec *telemetry.Record) {
+	if p.ctx.Err() != nil {
+		return
+	}
+
+	// Alerts and errors are not published to MQTT
+	if rec.TxType == "alerts" || rec.TxType == "errors" {
+		p.logger.Log(logrus.DEBUG, "mqtt_ignored_record", logrus.LogInfo{"tx_type": rec.TxType})
+		return
+	}
+
+	payload, err := rec.GetProtoMessage()
+	if err != nil {
+		p.ReportError("mqtt_payload_unmarshal_error", err, nil)
+		return
+	}
+
+	data, ok := payload.(*protos.Payload)
+	if !ok {
+		p.ReportError("mqtt_invalid_payload_type", fmt.Errorf("expected *protos.Payload, got %T", payload), nil)
+		return
+	}
+
+	convertedPayload := transformers.PayloadToMap(data, false, p.logger)
+
+	var tokens []pahomqtt.Token
+
+	// Publish each field to the relevant topic on the MQTT broker
+	for key, value := range convertedPayload {
+		// Skip VIN and CreatedAt fields
+		if key == "Vin" || key == "CreatedAt" {
+			continue
+		}
+
+		topicName := fmt.Sprintf("%s/vin/%s/%s", p.config.TopicBase, rec.Vin, key)
+
+		jsonValue, err := json.Marshal(map[string]interface{}{"value": value})
+		if err != nil {
+			p.ReportError("mqtt_json_marshal_error", err, logrus.LogInfo{"key": key})
+			continue
+		}
+
+		token := p.client.Publish(topicName, p.config.QoS, p.config.Retained, jsonValue)
+		tokens = append(tokens, token)
+
+		metricsRegistry.byteTotal.Add(int64(len(jsonValue)), map[string]string{"record_type": rec.TxType})
+		metricsRegistry.publishCount.Inc(map[string]string{"record_type": rec.TxType})
+	}
+
+	var publishError bool
+	// Wait for all topics to be published
+	for _, token := range tokens {
+		if !token.WaitTimeout(time.Duration(p.config.PublishTimeout) * time.Second) {
+			metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+			p.ReportError("mqtt_publish_timeout", fmt.Errorf("publish operation timed out"), logrus.LogInfo{"topic": "multiple"})
```

Comment:
logging txid and vin will help here in better debugging

Reply:
I don't think any other part of the existing code does that? And, so far, all txid's I have seen are empty strings. So, could change it, but I doubt if it is needed.

Reply:
Ok, I had a better look. Agreed, I will implement it.